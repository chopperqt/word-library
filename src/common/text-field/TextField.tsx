import { ChangeEvent, useEffect } from 'react'
import { useController } from "react-hook-form"

import Input from "components/Input"
import {
  FIELD_REQUIRED_TEXT,
  UNACCEPTABLE_SYMBOL_TEXT,
} from 'helpers/texts'

export interface TextFieldProps {
  name: string
  control?: any
  placeholder?: string
  value?: string
  isRequired?: boolean
}
const TextField = ({
  name,
  placeholder,
  control,
  value,
  isRequired = false,
}: TextFieldProps) => {
  const {
    field: {
      onChange: fieldUpdate,
      value: fieldValue,
      ref,
      onBlur,
    },
    fieldState: {
      error,
    },
  } = useController({
    control,
    name,
    rules: {
      required: {
        value: isRequired,
        message: FIELD_REQUIRED_TEXT,
      },
      pattern: {
        value: /^[a-zA-Z\s]+$/,
        message: UNACCEPTABLE_SYMBOL_TEXT,
      }
    },
  })

  const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
    fieldUpdate(e.target.value)
  }

  useEffect(() => {
    if (!value) return

    fieldUpdate(value)
  }, [value])

  return (
    <div className='h-14'>
      <Input
        data-testid="text-field"
        ref={ref}
        onBlur={onBlur}
        value={fieldValue}
        placeholder={placeholder}
        name={name}
        onChange={handleChange}
      />
      {error?.message && (
        <div
          className='text-sm text-rose-700'
          data-testid="text-field-error"
        >
          {error.message}
        </div>
      )}
    </div>
  )
}

export default TextField
