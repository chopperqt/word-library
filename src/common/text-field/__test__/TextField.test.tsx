import {
  render,
  screen,
  cleanup,
  fireEvent,
  act,
} from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'

import TextField from '../TextField'
import { useForm } from 'react-hook-form'

afterEach(cleanup)

it('Test render correctly', () => {
  const Component = () => {
    const {
      control,
    } = useForm({
      mode: 'onChange',
    })

    return (
      <TextField
        name="name"
        control={control}
      />
    )
  }

  render(<Component />)
})

describe('Test value', () => {
  const Component = () => {
    const {
      control,
    } = useForm({
      mode: 'onChange',
    })

    return (
      <TextField
        name="name"
        value='test default value'
        control={control}
      />
    )
  }

  it('Test default value', () => {
    render(<Component />)

    expect(screen.getByTestId('text-field')).toHaveValue('test default value')
  })

  it('Test change value', () => {
    render(<Component />)

    const input = screen.getByTestId('text-field')

    fireEvent.change(input, {
      target: {
        value: '',
      }
    })

    expect(input).toHaveValue("")

    fireEvent.change(input, {
      target: {
        value: 'second value change',
      }
    })

    expect(input).toHaveValue('second value change')
  })
})

describe('Test display errors', () => {
  const Component = () => {
    const {
      control,
    } = useForm({
      mode: 'onChange',
    })

    return (
      <TextField
        name="name"
        value='some text'
        control={control}
        isRequired={true}
        words={['test']}
        isCheckUniqueWord={true}
      />
    )
  }

  it('Check required error message', async () => {
    render(<Component />)

    const input = screen.getByTestId('text-field')

    fireEvent.change(input, {
      target: {
        value: '',
      }
    })

    const error = await screen.findByText('Field is required!')

    expect(error).toMatchSnapshot()
  })

  it('Check already exist error message', async () => {
    render(<Component />)

    const input = screen.getByTestId('text-field')

    fireEvent.change(input, {
      target: {
        value: 'test',
      }
    })

    expect(input).toHaveValue('test')

    const error = await screen.findByTestId('text-field-error')

    expect(error).toMatchSnapshot()
  })
})
