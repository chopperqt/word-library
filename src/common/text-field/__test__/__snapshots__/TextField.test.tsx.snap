// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test display errors Check already exist error message 1`] = `
<div
  class="text-sm text-rose-700 flex"
  data-testid="text-field-error"
>
  The word already exists
</div>
`;

exports[`Test display errors Check required error message 1`] = `
<div
  class="text-sm text-rose-700 flex"
  data-testid="text-field-error"
>
  Field is required!
</div>
`;
